File: app/src/main/java/com/savory/api/SavoryClient.java
package com.savory.api;

import android.support.annotation.NonNull;

import com.savory.api.models.AccountInfo;
import com.savory.api.models.Photo;
import com.savory.api.models.SavoryToken;

import java.util.List;

import retrofit2.Call;

public class SavoryClient {

    private static final String BASE_URL = "http://savory-backend.herokuapp.com";

    private static SavoryClient instance;

    private SavoryService savoryService;

    public SavoryClient() {

        savoryService = RetrofitBuilderFactory.createBaseRetrofit()
            .baseUrl(BASE_URL)
            .build()
            .create(SavoryService.class);
    }

    public static SavoryClient get() {
        if (instance == null) {
            synchronized (SavoryClient.class) {
                if (instance == null) {
                    instance = new SavoryClient();
                }
            }
        }

        return instance;
    }

    public Call<SavoryToken> connect(String socialPlatformToken) {
        return savoryService.connectWithSocial(socialPlatformToken);
    }

    public Call<AccountInfo> getMyAccountInfo(@NonNull String savoryToken) {
        return savoryService.getMyAccountInfo(savoryToken);
    }

    public Call<List<Photo>> getMyPhotos(@NonNull String savoryToken,
                                         int lastId) {
        return savoryService.getMyPhotos(savoryToken, lastId);
    }
}

File: app/src/main/java/com/savory/api/SavoryClient.kt
package com.savory.api

import com.savory.api.models.SavoryToken
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Call
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory

class SavoryClient {

    private val savoryService : SavoryService
    private val retrofit : Retrofit

    init {
        val httpLoggingInterceptor = HttpLoggingInterceptor();
        httpLoggingInterceptor.level = HttpLoggingInterceptor.Level.BODY

        val client = OkHttpClient.Builder()
                .addInterceptor(httpLoggingInterceptor)
                .build()

        retrofit = Retrofit.Builder()
                .client(client)
                .baseUrl(BASE_URL)
                .addConverterFactory(MoshiConverterFactory.create())
                .build()

        savoryService = retrofit.create(SavoryService::class.java)
    }

    fun connect(socialPlatformToken : String) : Call<SavoryToken> {
        return savoryService.connectWithSocial(socialPlatformToken)
    }

    companion object {
        const val BASE_URL = "http://savory-backend.herokuapp.com"

        val instance = SavoryClient()

        fun get() : SavoryClient {
            return instance
        }
    }
}